dds/DCPS/XTypes/DynamicData.h

```diff
@@ -0,0 +1,277 @@
+/*
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#ifndef OPENDDS_DCPS_XTYPES_DYNAMIC_DATA_H
+#define OPENDDS_DCPS_XTYPES_DYNAMIC_DATA_H
+
+#include "TypeObject.h"
+#include "DynamicType.h"
+
+#include <dds/DCPS/Serializer.h>
+#include <dds/DCPS/PoolAllocator.h>
+#include <dds/CorbaSeq/LongSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/ULongSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/Int8SeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/UInt8SeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/ShortSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/UShortSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/LongLongSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/ULongLongSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/FloatSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/DoubleSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/LongDoubleSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/CharSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/WCharSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/OctetSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/BooleanSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/StringSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/WStringSeqTypeSupportImpl.h>
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace XTypes {
+
+class OpenDDS_Dcps_Export DynamicData {
+public:
+  DynamicData();
+
+  /// This creates a duplicated ACE_Message_Block chain from the provided chain.
+  /// The duplicated chain is released when the object is destroyed. Caller is
+  /// responsible for the release of the input message block chain.
+  DynamicData(ACE_Message_Block* chain,
+              const DCPS::Encoding& encoding,
+              const DynamicType_rch& type);
+  DynamicData(DCPS::Serializer& ser, const DynamicType_rch& type);
+  DynamicData& operator=(const DynamicData& other);
+
+  ~DynamicData();
+
+  DDS::ReturnCode_t get_descriptor(MemberDescriptor& value, MemberId id) const;
+  DDS::ReturnCode_t set_descriptor(MemberId id, const MemberDescriptor& value);
+
+  MemberId get_member_id_by_name(DCPS::String name) const;
+  MemberId get_member_id_at_index(ACE_CDR::ULong index);
+  ACE_CDR::ULong get_item_count();
+
+  DynamicData clone() const;
+
+  DDS::ReturnCode_t get_int32_value(ACE_CDR::Long& value, MemberId id);
```

Comment:
Should this whole group (both `get_*_value` and `get_*_values`) be const members?

Reply:
Unfortunately these functions need to update the associated Serializer object, which is a member of the class. So I can't make them const members.

Reply:
I don't think we need to do anything for it in this PR, but it may be good to revisit later.  `ACE_Message_Block::duplicate()` is const and Serializers can each be local to the method that needs them.

Reply:
Right. We can remove the `strm_` data member and the code would be cleaner. I guess I was just avoiding passing the Serializer object from `get_*_value` or `get_*_values` to the other methods that they use. I'll revisit this.